### PR TITLE
refactor: remove `Validation` from `Typer`

### PR DESCRIPTION
**Pull request by jaschdoc**
Related to https://github.com/flix/flix/pull/8944
Related to https://github.com/flix/flix/issues/8815

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver.scala [hunk 1]
```diff
@@ -108,11 +108,11 @@ object ConstraintSolver {
           // If there are any constraints we could not resolve, then we report an error.
           // TODO ASSOC-TYPES here we only consider the first error
           getFirstError(deferred, renv) match {
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver.scala:110]: I can now change this to report all errors if we want to?

**Comment by jaschdoc** [comment]
We can do a1ff7a4771b8f2b11b0636cc8f4f0157e5c7185e in a separate PR if we want to

File: main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver.scala]: I think we should. Agree @mlutze ?

File: main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver.scala [hunk 2]
```diff
@@ -480,13 +476,12 @@ object ConstraintSolver {
...
   /**
-    * Gets an error from the list of unresolved constraints.
+    * Returns the errors from the list of unresolved constraints.
     */
-  private def getFirstError(deferred: List[TypeConstraint], renv: RigidityEnv)(implicit flix: Flix): Option[TypeError] = deferred match {
-    case Nil => None
-    case TypeConstraint.Equality(tpe1, tpe2, prov) :: _ => Some(toTypeError(UnificationError.MismatchedTypes(tpe1, tpe2), prov))
-    case TypeConstraint.Trait(sym, tpe, loc) :: _ => Some(mkMissingInstance(sym, tpe, renv, loc))
-    case TypeConstraint.Purification(_, _, _, _, nested) :: _ => getFirstError(nested, renv)
+  private def getErrorsFromTypeConstraints(deferred: List[TypeConstraint], renv: RigidityEnv)(implicit flix: Flix): List[TypeError] = deferred.flatMap {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver.scala:481]: Hey! This isn't a refactor! :O 

Let's make this change separately. There might be unexpected consequences (like having 999999 errors)

Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver.scala:481]: Yeah, that's also why I suggested making this change in a separate PR

Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver.scala:481]: But just wanted to show that it was possible now

**Comment by magnus-madsen** [comment]
@mlutze Let me know when you are happy then I will take a look.

**Comment by mlutze** [comment]
LGTM

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
One step closer!

**Pull request merged**